doc/transactions.md

```diff
@@ -0,0 +1,91 @@
+```
+
+The corresponding code in C would be:
+
+```C
+DLiteInstance *A = dlite_instance_get("A");  // Initial state (Fig. 1a)
+dlite_instance_snapshot(A);                  // Make snapshot at time=t1 (Fig. 1b)
+// A is evolved...
+dlite_instance_snapshot(A);                  // Make snapshot at time=t2 (Fig. 1c)
+```
+
+The snapshots can be accessed using `dlite_instance_get_snapshot()`.
+For instance, accessing snapshot `A2` and creating a new branch from it can be achieved by the following four lines of Python code:
+
+```python
+A2 = A.get_snapshot()  # Access the most recent snapshot of A
+B = A2.copy()          # Create a mutable copy of A2
+B.set_parent(A2)       # Make B a transaction with A2 as parent. Shown in Fig. 2a.
+B.shapshot()           # Make a snapshop of B. Shown in Fig. 2b.
+```
+
+The corresponding C code is:
+
+```C
+// Access the most recent snapshot of instance A.
+// Note that A2 is a borrowed reference to the snapshot and should not be
+// dereferenced with dlite_instance_decref().
+const DLiteInstance *A2 = dlite_instance_get_snapshot(A, 1);
+
+// Create a mutable copy of A2.
+// Note that B is a new reference.  You should call dlite_instance_decref() when you
+// are done with it.
+DLiteInstance *B = dlite_instance_copy(A2);
+
+// Make B a transaction with A2 as parent. Shown in Fig. 2a.
+dlite_instance_set_parent(B, A2);
+
+// Make a snapshop of B. Shown in Fig. 2b.
+dlite_instance_snapshot(B);
+```
+The result of these commands are shown in Figure 2.
+
+![transactions-branch](figs/transactions-branch.png)
+
+**Figure 2**: *Creating a new branch, from a snapshot of a transaction.
+(a) Create a copy (`B`) of latest snapshot of transaction `A`.
+(b) Take a snapshot of `B`.*
+
+### Transaction parent and immutability
+
+All transactions start as a root instance with no parent instance.
+All other instances in a transaction has exactly **one** parent instance.
+All instances in a transaction that serves as a parent are immutable (that is, all instances except the leaves).
+Non-root transaction instances store a [SHA-3](https://en.wikipedia.org/wiki/SHA-3) hash of their parent together with the parent UUID.
+This make it possible to ensure that any of the ancestors of a transaction has not been changed - providing provenance.
+
+A transaction can be validated with the `verify_transaction()` method in Python and `dlite_instance_verify_transaction()` in C.
+
+### Memory management
+
+The number of snapshots can potentially be very large, hence it is important to be able to store them to disk in order to save memory.
+To support this, DLite implements the `pull_snapshot()` and `push_snapshot()` methods (`dlite_instance_pull_snapshot()` and `dlite_instance_push_snapshot()` in C).
+
+Note that not all storages can be used with these functions, since whether an instance is a transactions or not, is not described by its metadata and hence requires special support by the storage plugin.
```

Comment:
It might be nice to write a bit more about the future of memory management, specifically how it's stored on disk to save space and such? It doesn't need to be much, just some empty promises that similar to `git`, we might try to pack/bundle old snapshots together to save even more space on disk?